File: bindings/lua/src/asset.lua
local AssetManager = {}
AssetManager.__index = AssetManager

-- Constants
local MAX_ASSET_NAME_LENGTH = 64
local MAX_ASSET_DESCRIPTION_LENGTH = 1024
local MAX_ASSET_METADATA_SIZE = 1024 * 1024  -- 1MB
local SUPPORTED_ASSET_TYPES = {
    "fungible",
    "non_fungible",
    "semi_fungible"
}

function AssetManager.new()
    local self = setmetatable({}, AssetManager)
    self.initialized = false
    self.blockchainManager = nil
    self.quantumManager = nil
    self.assets = {}
    return self
end

function AssetManager:initialize(blockchainManager, quantumManager)
    if self.initialized then
        error("AssetManager already initialized")
    end

    if not blockchainManager then
        error("BlockchainManager is required")
    end

    if not quantumManager then
        error("QuantumManager is required")
    end

    self.blockchainManager = blockchainManager
    self.quantumManager = quantumManager
    self.initialized = true
end

function AssetManager:createAsset(name, type, description, metadata, owner)
    if not self.initialized then
        error("AssetManager not initialized")
    end

    -- Validate parameters
    if not name or #name > MAX_ASSET_NAME_LENGTH then
        error("Invalid asset name")
    end

    if not type or not self:_isValidAssetType(type) then
        error("Invalid asset type")
    end

    if description and #description > MAX_ASSET_DESCRIPTION_LENGTH then
        error("Description too long")
    end

    if metadata and #metadata > MAX_ASSET_METADATA_SIZE then
        error("Metadata too large")
    end

    if not owner then
        error("Owner address is required")
    end

    -- Generate asset ID
    local assetId = self:_generateAssetId(name, owner)

    -- Create asset
    local asset = {
        id = assetId,
        name = name,
        type = type,
        description = description or "",
        metadata = metadata or {},
        owner = owner,
        created_at = os.time(),
        supply = 0,
        decimals = 18,
        version = 1
    }

    -- Store asset
    self.assets[assetId] = asset

    return asset
end

function AssetManager:getAsset(assetId)
    if not self.initialized then
        error("AssetManager not initialized")
    end

    if not assetId then
        error("Asset ID is required")
    end

    return self.assets[assetId]
end

function AssetManager:updateAsset(assetId, updates)
    if not self.initialized then
        error("AssetManager not initialized")
    end

    if not assetId then
        error("Asset ID is required")
    end

    local asset = self.assets[assetId]
    if not asset then
        error("Asset not found")
    end

    -- Validate updates
    if updates.name and #updates.name > MAX_ASSET_NAME_LENGTH then
        error("Invalid asset name")
    end

    if updates.description and #updates.description > MAX_ASSET_DESCRIPTION_LENGTH then
        error("Description too long")
    end

    if updates.metadata and #updates.metadata > MAX_ASSET_METADATA_SIZE then
        error("Metadata too large")
    end

    -- Apply updates
    for key, value in pairs(updates) do
        if key ~= "id" and key ~= "created_at" and key ~= "version" then
            asset[key] = value
        end
    end

    asset.version = asset.version + 1

    return asset
end

function AssetManager:mintAsset(assetId, amount, recipient)
    if not self.initialized then
        error("AssetManager not initialized")
    end

    if not assetId then
        error("Asset ID is required")
    end

    if not amount or amount <= 0 then
        error("Invalid amount")
    end

    if not recipient then
        error("Recipient address is required")
    end

    local asset = self.assets[assetId]
    if not asset then
        error("Asset not found")
    end

    -- Update supply
    asset.supply = asset.supply + amount

    -- Create mint transaction
    local transaction = {
        type = "mint",
        asset_id = assetId,
        amount = amount,
        recipient = recipient,
        timestamp = os.time()
    }

    -- Submit transaction to blockchain
    return self.blockchainManager:addTransaction(transaction)
end

function AssetManager:burnAsset(assetId, amount, owner)
    if not self.initialized then
        error("AssetManager not initialized")
    end

    if not assetId then
        error("Asset ID is required")
    end

    if not amount or amount <= 0 then
        error("Invalid amount")
    end

    if not owner then
        error("Owner address is required")
    end

    local asset = self.assets[assetId]
    if not asset then
        error("Asset not found")
    end

    if asset.owner ~= owner then
        error("Unauthorized")
    end

    if asset.supply < amount then
        error("Insufficient supply")
    end

    -- Update supply
    asset.supply = asset.supply - amount

    -- Create burn transaction
    local transaction = {
        type = "burn",
        asset_id = assetId,
        amount = amount,
        owner = owner,
        timestamp = os.time()
    }

    -- Submit transaction to blockchain
    return self.blockchainManager:addTransaction(transaction)
end

function AssetManager:transferAsset(assetId, amount, from, to)
    if not self.initialized then
        error("AssetManager not initialized")
    end

    if not assetId then
        error("Asset ID is required")
    end

    if not amount or amount <= 0 then
        error("Invalid amount")
    end

    if not from or not to then
        error("From and to addresses are required")
    end

    local asset = self.assets[assetId]
    if not asset then
        error("Asset not found")
    end

    -- Create transfer transaction
    local transaction = {
        type = "transfer",
        asset_id = assetId,
        amount = amount,
        from = from,
        to = to,
        timestamp = os.time()
    }

    -- Submit transaction to blockchain
    return self.blockchainManager:addTransaction(transaction)
end

function AssetManager:getAssetBalance(assetId, address)
    if not self.initialized then
        error("AssetManager not initialized")
    end

    if not assetId then
        error("Asset ID is required")
    end

    if not address then
        error("Address is required")
    end

    local asset = self.assets[assetId]
    if not asset then
        error("Asset not found")
    end

    -- Get balance from blockchain
    return self.blockchainManager:getBalance(assetId, address)
end

function AssetManager:_generateAssetId(name, owner)
    -- This is a placeholder for the actual implementation
    -- In a real implementation, this would use a proper hashing algorithm
    return "asset_" .. name .. "_" .. owner .. "_" .. os.time()
end

function AssetManager:_isValidAssetType(type)
    for _, validType in ipairs(SUPPORTED_ASSET_TYPES) do
        if type == validType then
            return true
        end
    end
    return false
end

return AssetManager 

File: bindings/typescript/src/asset.ts
import { satox_native } from './native/satox_native';
import { ThreadSafeManager } from './thread-safe';

export interface Asset {
  id: string;
  name: string;
  symbol: string;
  description: string;
  quantity: number;
  decimals: number;
  reissuable: boolean;
  owner: string;
  createdAt: number;
  updatedAt: number;
}

export interface AssetUpdate {
  name?: string;
  description?: string;
  quantity?: number;
}

export interface AssetTransfer {
  assetId: string;
  fromAddress: string;
  toAddress: string;
  quantity: number;
  fee: number;
}

export interface AssetHistory {
  assetId: string;
  fromAddress: string;
  toAddress: string;
  quantity: number;
  txHash: string;
  timestamp: number;
  type: number;
}

export interface AssetStats {
  totalAssets: number;
  totalTransfers: number;
  totalVolume: number;
  averageAssetValue: number;
  mostTradedAsset: string;
}

export class AssetManager extends ThreadSafeManager {
  private manager: any;
  private stats: AssetStats = {
    totalAssets: 0,
    totalTransfers: 0,
    totalVolume: 0,
    averageAssetValue: 0,
    mostTradedAsset: ''
  };

  constructor() {
    super();
    this.manager = new satox_native.AssetManager();
  }

  /**
   * Initialize the asset manager
   */
  async initialize(): Promise<boolean> {
    return this.withLock(async () => {
      try {
        this.ensureNotDisposed();
        const result = this.manager.initialize();
        this.setInitialized(result);
        this.setError(null);
        return result;
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        console.error('Failed to initialize Asset Manager:', error);
        return false;
      }
    });
  }

  /**
   * Shutdown the asset manager
   */
  async shutdown(): Promise<void> {
    return this.withLock(async () => {
      try {
        this.ensureNotDisposed();
        this.manager.shutdown();
        this.setInitialized(false);
        this.setError(null);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        console.error('Failed to shutdown Asset Manager:', error);
        throw error;
      }
    });
  }

  /**
   * Create a new asset
   */
  async createAsset(asset: Omit<Asset, 'id' | 'createdAt' | 'updatedAt'>): Promise<string> {
    return this.withLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.createAsset(asset);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return '';
      }
    });
  }

  /**
   * Get an asset by its ID
   */
  async getAsset(assetId: string): Promise<Asset | null> {
    return this.withReadLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.getAsset(assetId);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return null;
      }
    });
  }

  /**
   * Get asset metadata
   */
  async getAssetMetadata(assetId: string): Promise<Asset | null> {
    return this.withReadLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.getAssetMetadata(assetId);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return null;
      }
    });
  }

  /**
   * Update an asset
   */
  async updateAsset(assetId: string, update: AssetUpdate): Promise<boolean> {
    return this.withLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.updateAsset(assetId, update);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return false;
      }
    });
  }

  /**
   * Delete an asset
   */
  async deleteAsset(assetId: string): Promise<boolean> {
    return this.withLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.deleteAsset(assetId);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return false;
      }
    });
  }

  /**
   * List assets with pagination
   */
  async listAssets(limit: number = 100, offset: number = 0): Promise<Asset[]> {
    return this.withReadLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.listAssets(limit, offset);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return [];
      }
    });
  }

  /**
   * Search assets by query
   */
  async searchAssets(query: string, limit: number = 100): Promise<Asset[]> {
    return this.withReadLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.searchAssets(query, limit);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return [];
      }
    });
  }

  /**
   * Get asset balance for an address
   */
  async getAssetBalance(assetId: string, address: string): Promise<number> {
    return this.withReadLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.getAssetBalance(assetId, address);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return 0;
      }
    });
  }

  /**
   * Transfer assets between addresses
   */
  async transferAsset(transfer: AssetTransfer): Promise<string> {
    return this.withLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.transferAsset(transfer);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return '';
      }
    });
  }

  /**
   * Get asset transfer history
   */
  async getAssetHistory(assetId: string, limit: number = 100): Promise<AssetHistory[]> {
    return this.withReadLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.getAssetHistory(assetId, limit);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return [];
      }
    });
  }

  /**
   * Validate asset name
   */
  async validateAssetName(assetName: string): Promise<boolean> {
    return this.withReadLock(async () => {
      try {
        this.ensureInitialized();
        return this.manager.validateAssetName(assetName);
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return false;
      }
    });
  }

  /**
   * Get asset statistics
   */
  async getAssetStats(): Promise<AssetStats> {
    return this.withReadLock(async () => {
      try {
        this.ensureInitialized();
        return { ...this.stats };
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error';
        this.setError(errorMessage);
        return { ...this.stats };
      }
    });
  }

  /**
   * Check if the manager is initialized
   */
  isInitialized(): boolean {
    return this.isInitialized();
  }

  /**
   * Get the last error message
   */
  override getLastError(): string {
    return super.getLastError() || '';
  }

  /**
   * Clear the last error message
   */
  override async clearLastError(): Promise<void> {
    return this.withLock(async () => {
      super.clearLastError();
    });
  }
} 